mixer/pkg/cache/lruCache.go

```diff
@@ -113,25 +122,33 @@ func NewLRU(defaultExpiration time.Duration, evictionInterval time.Duration, max
 
 	if evictionInterval > 0 {
 		c.baseTimeNanos = time.Now().UTC().UnixNano()
-		c.evictionTicker = time.NewTicker(evictionInterval)
-		go c.evicter()
+		c.stopEvicter = make(chan bool, 1)
+		go c.evicter(evictionInterval)
+
+		// We return a 'see-through' wrapper for the real object such that
+		// the finalizer can trigger on the wrapper. We can't set a finalizer
+		// on the main cache object because it would never fire, because the
+		// evicter goroutine is keeping it alive
+		result := &lruWrapper{c}
+		runtime.SetFinalizer(result, func(w *lruWrapper) { c.stopEvicter <- true })
```

Comment:
whoa. i don't think i've ever looked at code with `runtime.SetFinalizer` before. just so I better understand the intent: this is just to make sure that any caches created will absolutely stop the created `Ticker` when they are no longer in use? Is this because we don't trust `Close()` ?

Reply:
A common pattern is to have an object that has some goroutines behind it. When you call Close, the goroutines are terminated cleanly. If you forget to call Close however, the goroutines are not terminated. Moreover, because the goroutines are there, they keep their "parent" object alive, and anything that this object refers to.

So in this case, if you have an LRU cache with 10000 slots in it and you forgot to call Close, you will effectively have a memory leak of 300K-400K. Not good.

A big part of the beauty of a GC environment is that you don't need to worry about the lifetime of objects. Requiring the use of Close methods erodes this benefit considerably. Close makes it so you can trivially cause unrecoverable memory leaks in your Go program. Not good.